shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -167,12 +169,21 @@ - (void)setViewController:(FlutterViewController*)viewController {
   _viewController = [viewController getWeakPtr];
   self.iosPlatformView->SetOwnerViewController(_viewController);
   [self maybeSetupPlatformViewChannels];
+
+  self.flutterViewControllerWillDeallocObserver =
+      [[NSNotificationCenter defaultCenter] addObserverForName:FlutterViewControllerWillDealloc
+                                                        object:viewController
+                                                         queue:[NSOperationQueue mainQueue]
+                                                    usingBlock:^(NSNotification* note) {
+                                                      [self notifyViewControllerDeallocated];
+                                                    }];
 }
 
 - (void)notifyViewControllerDeallocated {
   if (!_allowHeadlessExecution) {
     [self destroyContext];
   }
+  _viewController.reset();
 }
```

Comment:
This should only happen once right? Should we get rid of the observer here too?

Reply:
We could.  I wouldn't recommend it.  It would make more sense inside of the observer's block but it isn't necessary there either since the semantics of the message are that the object is dying, it shouldn't be dying twice.